docs/foreign-keys.md

```diff
@@ -0,0 +1,199 @@
+
+One operation that didn't make this list is **Resolution of null pointers**.
+Null pointers are a classic problem with non-relational data stores: if document
+A points to document B, and then document B is removed from the system, NoSQL
+databases offer no system for automatically resolving the null reference
+(whether by raising an error, setting a default, or simply storing a `NULL`
+value for the foreign key). In the example given above, we could imagine a user
+deleting an Event referenced by a Poster, and then the UUID stored in
+Poster.event would not reference a coherent entity in the database.
+
+Null pointers are indeed a problem with Record-to-Record foreign keys, but this
+problem is already endemic to Grout: if a field is removed from a schema during
+an upgrade to a new version (i.e. a definition of a new RecordSchema), Grout
+silently permits the field to remain but become unaccessible. From the reverse
+perspective, if a field is _added_ in the new schema, old records will not store
+values for that field, and so will not appear in the relevant filters.
+
+While not precisely a "null pointer," the issue is functionally the same: some data
+in the system is silently pointing to nothing. While null reference is a serious
+issue, it is not specific to foreign
+keys in this context, and it will need to be addressed through the user
+interface in a way that is common to all null reference problems in Grout.
+
+### Evaluating the requirements
+
+#### Lookups
+
+Since Record-to-Record foreign keys would store UUIDs (as strings) for the
+referent Record, lookups should be no different than for any other Grout field.
+
+
+#### Validation
+
+The [Python jsonschema package](https://python-jsonschema.readthedocs.io) has
+no built-in support for validating foreign keys, but it does expose
+a `RefResolver` class that it uses in the `validate` function for [resolving external
+references](https://python-jsonschema.readthedocs.io). It seems relatively
+straightforward to extend this resolver to check references by pinging the Grout
+API.
+
+- Pros:
+    - Easy integration with existing jsonschema validation
+    - Minimal changes to the application code
+
+- Cons:
+    - Extending a third-party package risks introducing technical debt
+
+#### Indexing
+
+PostgreSQL [supports indexing on jsonb
+fields](https://www.postgresql.org/docs/9.4/static/datatype-json.html#JSON-INDEXING)
+using GIN indexes. There are even three different degrees of flexibility with
+jsonb indexes:
+
+- Key-value index on all keys and values (the default: most flexible, least performant)
+- Key-value index with expression indexes, which will restrict the index to
+  a certain set of pre-defined key-value pairs (medium flexible, medium performant)
+- `jsonb_path_ops` indexes, which only support the containment operator `@>`
+  but are much faster (least flexible, most performant)
+
+Since we currently only use the containment operator in Grout anyway, I see no
```

Comment:
That's good to know! I'll flesh out this section a little bit with information about Django's implementation. I think it would make sense to use the default index type, and then consider using `jsonb_path_ops` if performance becomes a serious concern.